File: prismpath/kernel/parser.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Parse a workflow markdown file into a graph.

Format (human-authored, git-diffable):

    ---
    name: bugfix
    start: triage
    ---

    ## triage
    Understand the bug report and decide what to do next.
    -> implement: the bug is reproduced and the root cause is clear
    -> gather_info: it cannot be reproduced or more information is needed
    -> close: it is a duplicate or invalid

    ## done
    Summarize and finish.

Each `## heading` is a node; the prose under it is the node's instruction. Lines of the form
`-> target: condition` are outgoing edges (natural-language conditions). A node with no edges
is terminal.
"""
from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from typing import Dict, List, Tuple

EDGE_RE = re.compile(r"^\s*-?\s*->\s*([A-Za-z0-9_\-]+)\s*:\s*(.+?)\s*$")
HEAD_RE = re.compile(r"^\s*##\s+(.+?)\s*$")
# Node annotation, e.g. `@checkpoint(unit=alert.id, proof=verdict, gate=staged_ok)` — an extension
# slot on a node (the control plane reads it; the kernel just parses it). Keeps the heading clean.
ANNO_RE = re.compile(r"^\s*@(\w+)\s*\((.*)\)\s*$")

# Input bounds. A flow is human-authored routing config, not a data payload: real ones are a handful
# of nodes and a few KiB. These caps keep an untrusted or accidental oversized document from
# exhausting memory or driving the polynomial static analysis (and the SCC recursion, whose limit
# scales with node count) off a cliff. All override via env, mirroring the Mission Control caps.
MAX_FLOW_BYTES = int(os.environ.get("PRISMPATH_MAX_FLOW_BYTES", str(2 * 1024 * 1024)))  # 2 MiB
MAX_NODES = int(os.environ.get("PRISMPATH_MAX_NODES", "5000"))
MAX_EDGES = int(os.environ.get("PRISMPATH_MAX_EDGES", "20000"))


class ParseError(ValueError):
    """A flow document could not be parsed within its input bounds. Subclasses ValueError so callers
    that already funnel bad input to a client error (e.g. Mission Control's 400 envelope) route it
    without special-casing."""


def _parse_anno_args(argstr: str) -> dict:
    """`key=value` pairs (e.g. @checkpoint(unit=alert.id)) AND bare tokens mapped to None (e.g.
    @emits(recommended_action, rule_level)). Backward-compatible: existing key=value annotations are
    unchanged."""
    args = {}
    for part in argstr.split(","):
        part = part.strip()
        if not part:
            continue
        if "=" in part:
            key, val = part.split("=", 1)
            key = key.strip()
            if key:                       # ignore a malformed `=value` with no key
                args[key] = val.strip()
        else:
            args[part] = None
    return args


@dataclass
class Node:
    name: str
    instruction: str = ""
    edges: List[Tuple[str, str]] = field(default_factory=list)  # (target, condition)
    annotations: Dict[str, dict] = field(default_factory=dict)  # {name: {arg: value}}

    @property
    def terminal(self) -> bool:
        return len(self.edges) == 0


@dataclass
class Graph:
    name: str
    start: str
    nodes: Dict[str, Node]
    meta: Dict[str, str] = field(default_factory=dict)  # frontmatter beyond name/start (e.g. packing)



def parse(text: str) -> Graph:
    # Bound before doing any work. len(text) is a character count — a tight lower proxy for bytes
    # (>= 1 byte/char) and the figure that actually governs the in-memory string and the line scan.
    if len(text) > MAX_FLOW_BYTES:
        raise ParseError(f"flow document exceeds PRISMPATH_MAX_FLOW_BYTES ({MAX_FLOW_BYTES} bytes)")
    meta = {}
    body = text
    match = re.match(r"^---\s*\n(.*?)\n---\s*\n(.*)$", text, re.DOTALL)
    if match:
        for line in match.group(1).splitlines():
            if ":" in line:
                key, val = line.split(":", 1)
                meta[key.strip()] = val.strip()
        body = match.group(2)

    nodes: Dict[str, Node] = {}
    cur: Node | None = None
    instr_lines: List[str] = []
    edge_count = 0

    def flush():
        if cur is not None:
            cur.instruction = "\n".join(instr_lines).strip()

    for line in body.splitlines():
        head_match = HEAD_RE.match(line)
        if head_match:
            flush()
            name = head_match.group(1).strip().lower().replace(" ", "_")
            # Count distinct node names — a new heading that reuses a name isn't a new node.
            if name not in nodes and len(nodes) >= MAX_NODES:
                raise ParseError(f"flow document exceeds PRISMPATH_MAX_NODES ({MAX_NODES} nodes)")
            cur = Node(name=name)
            nodes[name] = cur
            instr_lines = []
            continue
        if cur is None:
            continue
        edge_match = EDGE_RE.match(line)
        anno_match = ANNO_RE.match(line)
        if edge_match:
            edge_count += 1
            if edge_count > MAX_EDGES:
                raise ParseError(f"flow document exceeds PRISMPATH_MAX_EDGES ({MAX_EDGES} edges)")
            cur.edges.append((edge_match.group(1).strip(), edge_match.group(2).strip()))
        elif anno_match:
            # merge repeated annotations of the same name (e.g. @emits split across two lines) rather
            # than overwriting, so a later declaration doesn't silently drop the earlier fields.
            cur.annotations.setdefault(anno_match.group(1).strip(), {}).update(_parse_anno_args(anno_match.group(2)))
        else:
            instr_lines.append(line)
    flush()

    start = meta.get("start") or (next(iter(nodes)) if nodes else "")
    parsed_graph = Graph(name=meta.get("name", "flow"), start=start, nodes=nodes, meta=meta)
    return parsed_graph


def parse_file(path: str) -> Graph:
    # Gate on the on-disk size before reading the file into memory — a huge file never gets slurped.
    sz = os.path.getsize(path)
    if sz > MAX_FLOW_BYTES:
        raise ParseError(
            f"flow file {os.path.basename(path)!r} is {sz} bytes, exceeds "
            f"PRISMPATH_MAX_FLOW_BYTES ({MAX_FLOW_BYTES})")
    with open(path, encoding="utf-8") as file_handle:
        return parse(file_handle.read())


def reachable(graph) -> set:
    """Node names reachable from the start node by any edge (graph traversal only, no predicate
    evaluation). Shared by the analyzer, the Level M classifier, and the model checker."""
    seen, stack = set(), [graph.start]
    while stack:
        cur = stack.pop()
        if cur in seen or cur not in graph.nodes:
            continue
        seen.add(cur)
        for target, _cond in graph.nodes[cur].edges:
            stack.append(target)
    return seen

File: prismpath-rs/src/parser.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Markdown flow parser.

use std::collections::HashMap;

use crate::condition::{is_js_ws, js_trim, split_lines_py};

#[derive(Debug, Clone, Default)]
pub struct Node {
    pub name: String,
    pub instruction: String,
    pub edges: Vec<(String, String)>,
    pub annotations: HashMap<String, HashMap<String, Option<String>>>,
}

#[derive(Debug, Clone, Default)]
pub struct Graph {
    pub name: String,
    pub start: String,
    pub nodes: HashMap<String, Node>,
}

/// -> target: condition (optionally as a - bullet): EDGE_RE.
fn match_edge(line: &str) -> Option<(String, String)> {
    let line_chars: Vec<char> = line.chars().collect();
    let mut pos = 0;
    while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
        pos += 1;
    }
    // the optional bullet -
    if pos < line_chars.len() && line_chars[pos] == '-' && !(pos + 1 < line_chars.len() && line_chars[pos + 1] == '>') {
        pos += 1;
        while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
            pos += 1;
        }
    }
    if !(pos + 1 < line_chars.len() && line_chars[pos] == '-' && line_chars[pos + 1] == '>') {
        return None;
    }
    pos += 2;
    while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
        pos += 1;
    }
    let target_start = pos;
    while pos < line_chars.len() && (line_chars[pos].is_ascii_alphanumeric() || line_chars[pos] == '_' || line_chars[pos] == '-') {
        pos += 1;
    }
    if pos == target_start {
        return None;
    }
    let target: String = line_chars[target_start..pos].iter().collect();
    while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
        pos += 1;
    }
    if pos >= line_chars.len() || line_chars[pos] != ':' {
        return None;
    }
    pos += 1;
    let rest: String = line_chars[pos..].iter().collect();
    if rest.is_empty() {
        return None;
    }
    Some((target, js_trim(&rest).to_string()))
}

/// ## heading: HEAD_RE
fn match_head(line: &str) -> Option<String> {
    let line_chars: Vec<char> = line.chars().collect();
    let mut pos = 0;
    while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
        pos += 1;
    }
    if !(pos + 1 < line_chars.len() && line_chars[pos] == '#' && line_chars[pos + 1] == '#') {
        return None;
    }
    pos += 2;
    if pos >= line_chars.len() || !is_js_ws(line_chars[pos]) {
        return None;
    }
    if pos + 1 >= line_chars.len() {
        return None;
    }
    let rest: String = line_chars[pos + 1..].iter().collect();
    Some(js_trim(&rest).to_string())
}

/// @name(args): ANNO_RE
fn match_anno(line: &str) -> Option<(String, String)> {
    let line_chars: Vec<char> = line.chars().collect();
    let mut pos = 0;
    while pos < line_chars.len() && is_js_ws(line_chars[pos]) {
        pos += 1;
    }
    if pos >= line_chars.len() || line_chars[pos] != '@' {
        return None;
    }
    pos += 1;
    let name_start = pos;
    while pos < line_chars.len() && (line_chars[pos].is_alphanumeric() || line_chars[pos] == '_') {
        pos += 1;
    }
    if pos == name_start || pos >= line_chars.len() || line_chars[pos] != '(' {
        return None;
    }
    let name: String = line_chars[name_start..pos].iter().collect();
    let mut close_pos = None;
    for scan in (pos + 1..line_chars.len()).rev() {
        if line_chars[scan] == ')' && line_chars[scan + 1..].iter().all(|ch| is_js_ws(*ch)) {
            close_pos = Some(scan);
            break;
        }
    }
    let close_pos = close_pos?;
    let args: String = line_chars[pos + 1..close_pos].iter().collect();
    Some((name, args))
}

fn parse_anno_args(argstr: &str) -> Vec<(String, Option<String>)> {
    let mut args = Vec::new();
    for part in argstr.split(',') {
        let part_trimmed = js_trim(part);
        if part_trimmed.is_empty() {
            continue;
        }
        match part_trimmed.find('=') {
            Some(equals_pos) => {
                let key = js_trim(&part_trimmed[..equals_pos]);
                if !key.is_empty() {
                    args.push((key.to_string(), Some(js_trim(&part_trimmed[equals_pos + 1..]).to_string())));
                }
            }
            None => args.push((part_trimmed.to_string(), None)),
        }
    }
    args
}

/// Split off YAML frontmatter.
fn split_frontmatter(text: &str) -> Option<(String, String)> {
    let text_chars: Vec<char> = text.chars().collect();
    if text_chars.len() < 3 || text_chars[0] != '-' || text_chars[1] != '-' || text_chars[2] != '-' {
        return None;
    }
    let mut pos = 3;
    let mut last_newline = None;
    while pos < text_chars.len() && is_js_ws(text_chars[pos]) {
        if text_chars[pos] == '\n' {
            last_newline = Some(pos);
        }
        pos += 1;
    }
    let body_start = last_newline? + 1;
    let mut scan = body_start;
    while scan + 3 < text_chars.len() {
        if text_chars[scan] == '\n' && text_chars[scan + 1] == '-' && text_chars[scan + 2] == '-' && text_chars[scan + 3] == '-' {
            let mut lookahead = scan + 4;
            let mut newline2 = None;
            while lookahead < text_chars.len() && is_js_ws(text_chars[lookahead]) {
                if text_chars[lookahead] == '\n' {
                    newline2 = Some(lookahead);
                }
                lookahead += 1;
            }
            if let Some(newline2) = newline2 {
                let meta: String = text_chars[body_start..scan].iter().collect();
                let rest: String = text_chars[newline2 + 1..].iter().collect();
                return Some((meta, rest));
            }
        }
        scan += 1;
    }
    None
}

/// Markdown -> Graph: a faithful port of the reference parse. A node with no edges is terminal.
pub fn parse(text: &str) -> Graph {
    let mut meta: HashMap<String, String> = HashMap::new();
    let body = match split_frontmatter(text) {
        Some((frontmatter, body_text)) => {
            for line in split_lines_py(&frontmatter) {
                if let Some(colon_pos) = line.find(':') {
                    meta.insert(
                        js_trim(&line[..colon_pos]).to_string(),
                        js_trim(&line[colon_pos + 1..]).to_string(),
                    );
                }
            }
            body_text
        }
        None => text.to_string(),
    };

    let mut nodes: HashMap<String, Node> = HashMap::new();
    let mut order: Vec<String> = Vec::new();
    let mut current_node: Option<String> = None;
    let mut instructions: Vec<String> = Vec::new();

    fn flush(
        cur_node: &Option<String>,
        instr_vec: &[String],
        nodes_map: &mut HashMap<String, Node>,
    ) {
        if let Some(name) = cur_node {
            if let Some(node) = nodes_map.get_mut(name) {
                node.instruction = js_trim(&instr_vec.join("\n")).to_string();
            }
        }
    }

    for line in split_lines_py(&body) {
        if let Some(heading_text) = match_head(line) {
            flush(&current_node, &instructions, &mut nodes);
            let name = heading_text.to_lowercase().replace(' ', "_");
            nodes.insert(name.clone(), Node { name: name.clone(), ..Default::default() });
            if !order.contains(&name) {
                order.push(name.clone());
            }
            current_node = Some(name);
            instructions = Vec::new();
            continue;
        }
        let Some(cur_name) = &current_node else { continue };
        if let Some((target, cond)) = match_edge(line) {
            nodes.get_mut(cur_name).expect("cur names an inserted node").edges.push((target, cond));
        } else if let Some((anno_name, argstr)) = match_anno(line) {
            let node = nodes.get_mut(cur_name).expect("cur names an inserted node");
            let entry = node.annotations.entry(anno_name).or_default();
            for (key, val) in parse_anno_args(&argstr) {
                entry.insert(key, val);
            }
        } else {
            instructions.push(line.to_string());
        }
    }
    flush(&current_node, &instructions, &mut nodes);

    let start = match meta.get("start") {
        Some(start_node) if !start_node.is_empty() => start_node.clone(),
        _ => order.first().cloned().unwrap_or_default(),
    };
    let name = meta.get("name").cloned().unwrap_or_else(|| "flow".to_string());
    Graph { name, start, nodes }
}
